File: mixin-bekasi-webface/src/main/js/bekasi/rest/RestFileRunspace.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import * as ajax from "webface/core/ajax";

import { jsonLeanFactory } from "webface/constants";

import DetailMessageDialog from "webface/dialogs/DetailMessageDialog";

import RunspaceItem from "bekasi/resources/RunspaceItem";
import RunspaceItemList from "bekasi/resources/RunspaceItemList";
import RunspaceHomeList from "bekasi/resources/RunspaceHomeList";

import RestSystemWorkspace from "bekasi/rest/RestSystemWorkspace";

export default class RestFileRunspace {

	private endPoint: string = null;

	constructor(endPoint: string) {
		this.endPoint = RestSystemWorkspace.WORKSPACE + endPoint;
	}

	public getHomeList(callback: (list: RunspaceHomeList) => void): void {
		ajax.doGet(this.endPoint + "", {
		}).done((json: any) => {
			let list = <RunspaceHomeList>jsonLeanFactory.create(json);
			callback(list);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Home List");
		});
	}

	public getItem(itemId: string, callback: (item: RunspaceItem) => void): void {
		ajax.doGet(this.endPoint + "/" + itemId, {
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Item");
		});
	}

	public removeItem(itemId: string, callback: () => void): void {
		ajax.doDelete(this.endPoint + "/" + itemId, {
		}).done(() => {
			callback();
		}).fail((error) => {
			DetailMessageDialog.open(error, "Remove Item");
		});
	}

	public getItemAncestors(itemId: string, callback: (items: RunspaceItem[]) => void): void {
		ajax.doGet(this.endPoint + "/" + itemId + "/ancestors", {
		}).done((jsons) => {
			let items: RunspaceItem[] = [];
			for (let i = 0; i < jsons.length; i++) {
				let json = jsons[i];
				let item = <RunspaceItem>jsonLeanFactory.create(json);
				items.push(item);
			}
			callback(items);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Item Ancestors");
		});
	}

	public getItemList(folderId: string, callback: (list: RunspaceItemList) => void): void {
		ajax.doGet(this.endPoint + "/" + folderId + "/list", {
		}).done((json: any) => {
			let list = <RunspaceItemList>jsonLeanFactory.create(json);
			callback(list);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Item List");
		});
	}

	public getItemNameList(folderId: string, callback: (list: string[]) => void): void {
		ajax.doGet(this.endPoint + "/" + folderId + "/names", {
		}).done((json) => {
			let list = <string[]>json;
			callback(list);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Item Name List");
		});
	}

	public postItemRename(itemId: string, newName: string, callback: (item: RunspaceItem) => void): void {
		ajax.doPost(this.endPoint + "/" + itemId + "/rename", {
			newName: newName,
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Post Item Rename");
		});
	}

	public postItemMove(itemId: string, newFolderId: string, callback: (item: RunspaceItem) => void): void {
		ajax.doPost(this.endPoint + "/" + itemId + "/move", {
			newFolderId: newFolderId,
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Post Item Move");
		});
	}

	public postItemCopy(itemId: string, newFolderId: string, newName: string,
		callback: (item: RunspaceItem) => void): void {
		ajax.doPost(this.endPoint + "/" + itemId + "/copy", {
			newFolderId: newFolderId,
			newName: newName,
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Post Item Copy");
		});
	}

	public postCreateFolder(itemId: string, name: string, callback: (item: RunspaceItem) => void): void {
		ajax.doPost(this.endPoint + "/" + itemId + "/create-folder", {
			name: name,
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Post Create Folder");
		});
	}

	public postCreateFile(itemId: string, name: string, callback: (item: RunspaceItem) => void): void {
		ajax.doPost(this.endPoint + "/" + itemId + "/create-file", {
			name: name,
		}).done((json) => {
			let item = <RunspaceItem>jsonLeanFactory.create(json);
			callback(item);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Post Create File");
		});
	}

	public getDependentFiles(fileId: string, callback: (fileIds: string[]) => void): void {
		ajax.doGet(this.endPoint + "/" + fileId + "/dependent-files", {
		}).done((fileIds: any[]) => {
			callback(fileIds);
		}).fail((error) => {
			DetailMessageDialog.open(error, "Get Dependent Files");
		});
	}

}

File: mixin-bekasi-webface/src/main/java/com/andia/mixin/bekasi/webface/RestFileRunspace.java
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.andia.mixin.bekasi.webface;

import static javax.ws.rs.core.MediaType.APPLICATION_FORM_URLENCODED;

import java.util.UUID;

import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.FormParam;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import com.andia.mixin.bekasi.Runserver;
import com.andia.mixin.bekasi.RunserverException;
import com.andia.mixin.bekasi.Runspace;
import com.andia.mixin.bekasi.resources.RunspaceItem;
import com.andia.mixin.bekasi.resources.RunspaceItemList;
import com.andia.mixin.lawang.SecuritySetting;

@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class RestFileRunspace {

	public static final String ANCESTORS = "ancestors";
	public static final String LIST = "list";
	public static final String RENAME = "rename";
	public static final String NAMES = "names";
	public static final String MOVE = "move";
	public static final String COPY = "copy";
	public static final String CREATE_FILE = "create-file";
	public static final String CREATE_FOLDER = "create-folder";
	public static final String DEPENDENT_FILES = "dependent-files";

	protected Runserver runserver;
	protected Runspace runspace;

	@Inject
	public void setSecuritySetting(SecuritySetting setting) throws RunserverException {
		String space = setting.getUserId();
		runspace = runserver.getRunspace(space);
	}

	@GET
	@Path("/")
	public RunspaceItemList getHomeList() throws Exception {
		RunspaceItemList list = runspace.getHomeList();
		return list;
	}

	@GET
	@Path("{itemId}")
	public RunspaceItem getItem(@PathParam("itemId") String itemId) throws Exception {
		UUID uuid = UUID.fromString(itemId);
		if (runspace.isItemExists(uuid)) {
			RunspaceItem item = runspace.getItem(uuid);
			return item;
		} else {
			throw new Exception("Missing item " + itemId);
		}
	}

	@DELETE
	@Path("{itemId}")
	public void removeItem(@PathParam("itemId") String itemId) throws Exception {
		UUID uuid = UUID.fromString(itemId);
		runspace.deleteItem(uuid);
	}

	@GET
	@Path("{itemId}/" + ANCESTORS)
	public RunspaceItem[] getAncestors(@PathParam("itemId") String itemId) throws Exception {
		UUID uuid = UUID.fromString(itemId);
		RunspaceItem[] items = runspace.getItemAncestors(uuid);
		return items;
	}

	@GET
	@Path("{folderId}/" + LIST)
	public RunspaceItemList listItems(@PathParam("folderId") String folderId) throws Exception {
		UUID uuid = UUID.fromString(folderId);
		RunspaceItemList list = runspace.getItemList(uuid);
		return list;
	}

	@GET
	@Path("{folderId}/" + NAMES)
	public String[] getItemNameList(@PathParam("folderId") String folderId) throws Exception {
		UUID uuid = UUID.fromString(folderId);
		String[] names = runspace.getItemNames(uuid);
		return names;
	}

	@POST
	@Path("{itemId}/" + RENAME)
	@Consumes(APPLICATION_FORM_URLENCODED)
	public RunspaceItem postItemRename(@PathParam("itemId") String itemId, @FormParam("newName") String newName)
			throws Exception {
		UUID uuid = UUID.fromString(itemId);
		RunspaceItem item = runspace.renameItem(uuid, newName);
		return item;
	}

	@POST
	@Path("{itemId}/" + MOVE)
	@Consumes(APPLICATION_FORM_URLENCODED)
	public RunspaceItem postItemMove(@PathParam("itemId") String itemId, @FormParam("newFolderId") String newFolderId)
			throws Exception {
		UUID itemUUID = UUID.fromString(itemId);
		UUID newFolderUUID = UUID.fromString(newFolderId);
		RunspaceItem item = runspace.moveItem(itemUUID, newFolderUUID);
		return item;
	}

	@POST
	@Path("{itemId}/" + COPY)
	@Consumes(APPLICATION_FORM_URLENCODED)
	public RunspaceItem postItemCopy(@PathParam("itemId") String itemId, @FormParam("newFolderId") String newFolderId,
			@FormParam("newName") String newName) throws Exception {
		UUID itemUUID = UUID.fromString(itemId);
		UUID newFolderUUID = UUID.fromString(newFolderId);
		RunspaceItem item = runspace.copyItem(itemUUID, newFolderUUID, newName);
		return item;
	}

	@POST
	@Path("{folderId}/" + CREATE_FOLDER)
	@Consumes(APPLICATION_FORM_URLENCODED)
	public RunspaceItem postCreateFolder(@PathParam("folderId") String folderId, @FormParam("name") String name)
			throws Exception {
		UUID uuid = UUID.fromString(folderId);
		RunspaceItem item = runspace.createFolder(uuid, name);
		return item;
	}

	@POST
	@Path("{folderId}/" + CREATE_FILE)
	@Consumes(APPLICATION_FORM_URLENCODED)
	public RunspaceItem postCreateFile(@PathParam("folderId") String folderId, @FormParam("name") String name)
			throws Exception {
		UUID uuid = UUID.fromString(folderId);
		RunspaceItem item = runspace.createFile(uuid, name);
		return item;
	}

	@GET
	@Path("{fileId}/" + DEPENDENT_FILES)
	public UUID[] getDependentFiles(@PathParam("fileId") String fileId) throws Exception {
		UUID uuid = UUID.fromString(fileId);
		UUID[] fileIds = runspace.getDependentFiles(uuid);
		return fileIds;
	}

}
